### DISCUSSION TITLE
Can't Query Content Type Created from Taxonomy Edit Page

### DISCUSSION THREAD

**Discussion by fuzl-llc** [OrchardCMS/OrchardCore#9612] [General]
I have a "Location" content type and a "Locations" taxonomy with the "Term Content Type" set to the "Location" content type. I also have a Lucene query for a "Location" content type:
 
Schema:
{
"type": "ContentItem/Location"
}
 
Query:
{
"query": {
"term": { "Content.ContentItem.ContentType": "Location" }
}
}
 
I add two content items: a location "Location 1" using the "Content > Content Items > New > Location" menu item and "Location 2" using the "Content > Content Items > Locations > Add Location" button (i.e. from within the "Edit Taxonomy" page of the "Locations" taxonomy). If I run the query above, it only shows "Location 1" even though they are both Location content types.
 
Any ideas on what the difference is between content items added directly vs those that are added from within a taxonomy that would make the query find/not find them? Can the query be modified to find those that were created from inside the taxonomy?

**Comment by Piedone** [comment]
Terms are only saved in the Taxonomy content item (in the item's Document), not as separate content items, and thus can't be queried like this. If you want to query content items that are connected to other content items you could e.g. use the Content Picker Field. If you want to use Taxonomies you can fetch the terms e.g. like this without server-side coding: https://docs.orchardcore.net/en/dev/docs/reference/modules/Taxonomies/#taxonomy_terms

**Reply by fuzl-llc** [reply]
Thanks, this is super helpful!  I think this will work for the case where I just want the full list of content items.  However, I have a second query where I want to query for a single Location content item based on a field defined on the Location content item:
```
{
  "query": {
    "bool" : {
    	"must" : {
          "term": { "ContentType": "Location" },
       		"term" : { "Location.Url" : "{{ url }}" }
        }
    }
  }
}
```
And I pass in the URL as a parameter.  With the taxonomy_terms filter I think I'll basically get a list of all (maybe hundreds of) locations.  I could loop through in a liquid template to get the one I want but that would be less then ideal.  Any way to just get the one I want from the server side?

**Reply by fuzl-llc** [reply]
I'm trying to adapt those referenced examples which have this:

Model.ContentItem.Content.BlogPost.Colors | taxonomy_terms

However in my case I'm trying to run this query from a Liquid page where I need to get one Location content type based on the URL found from {{ Request.Host }} - there's not actually a taxonomy field on the liquid page for the Locations taxonomy so not sure this can even work.

**Reply by fuzl-llc** [reply]
One workaround I've got working is to create a "Location Aware Liquid Page," put a "Locations" taxonomy field on it and select all terms in the taxonomy.  Then whenever my Location Aware Liquid Page loads it will already have all the locations in it and I can just use a liquid template to select the one I want.  This may even not be crazy if on those pages I want to show a menu with all the locations (which I probably do).  It might work for now but I'd still love to have a way to just query for a particular location by its Url field.

**Reply by fuzl-llc** [reply]
With a day's reflection, I have a better solution which is to simply create a Lucene query for the taxonomy Content Type itself.  I think I had to check a box to ensure that the Alias part would be included in the index (and of course add "Taxonomy" to the list of types to be indexed) but then I could create a simple query that got just the one taxonomy I needed (based on its alias "locations" in my case) with all contained terms and their fields.  Then from a liquid page it was easy to just loop through and grab the info I needed.  I'd still like to be able to JUST query for the one term I need based on the value of one of its fields but this is way better than the whole "Location Aware Liquid Page" idea.  Maybe getting just one term based on a field requires server side code...  if anyone has a way to do this I'd love to hear it.  No code would be great but a code solution would be welcome as well.  Thanks again @Piedone for the key to this all - that those terms are stored IN the taxonomy content item.

**Reply by fuzl-llc** [reply]
Also, if it disturbs anyone that you can create the taxonomy terms from two places, well, I also learned you can just turn that off using the "Creatable" setting for the taxonomy term type.  Then you'll never accidentally create one from there instead of from within the taxonomy's edit page.

**Reply by fuzl-llc** [reply]
And since I had a lot of trouble getting the query working I might as well include that here in case it might help someone:

Schema:
```
{
    "type": "ContentItem/Taxonomy"
}
```

Enable "Return Content Items"

Query:
```
{
  "query": {
    "term": { "Content.ContentItem.ContentType": "Taxonomy" },
    "term" : { "AliasPart" : "locations" }
  }
}
```

**Reply by Piedone** [reply]
Thanks for sharing the details!

**Comment by dpomt** [comment]
I was running exactly in the same issue and was wondering why some ContentItems of my Taxonomy TermContentType where indexed and some not.

I understand that the content items created within the taxonomy are only stored within the taxonomy content item and not as separate content items.
Anyway, I would like to use the "standard" OC search functionanlity to not only search for "normal" content items but also for the content items created within a taxonomy's hierarchy.
Is there any way how this could be achieved?

Thanks

**Reply by Piedone** [reply]
I think it should be. I won't be able to provide a complete solution now, but some pointers: you can specify what's actually indexed from a content item (or even content field) with content type settings. See https://docs.orchardcore.net/en/latest/docs/reference/modules/Media/#media-content-search for something similar. So, you could try something like this.

Also, I think those items should simply be indexed. What did you see as being indexed and what not?

**Comment by dpomt** [comment]
Thanks for your reply and suggestion.
I have previously done as suggested and included the fields to be indexed:

![image](https://github.com/OrchardCMS/OrchardCore/assets/1379660/7faa6dbe-18fa-46de-84c7-1ed4a4e6720c)

![image](https://github.com/OrchardCMS/OrchardCore/assets/1379660/7167d918-fefc-4255-8dc9-d7eac4bc3580)

![image](https://github.com/OrchardCMS/OrchardCore/assets/1379660/cc381f2e-5102-4069-a5d8-f1cb949446e6)

![image](https://github.com/OrchardCMS/OrchardCore/assets/1379660/a8bef099-42bb-49b1-9255-0c600bbb95d6)

I have edited and re-published tax terms. created new ones. But same behavor, the elements will not get indexed:

![image](https://github.com/OrchardCMS/OrchardCore/assets/1379660/53042c64-724e-4a0f-9383-a59c76dfb891)

The only term parts that are getting indexed (FEFCOTypenTerm...) are those I have not created within the taxonomy but directly via admin dashboard as "normal" content items.


**I have been debugging to see whether indexing is triggered.
It is not! context.ContentItem.Id is always 0 (even for newly created tax terms):**

![image](https://github.com/OrchardCMS/OrchardCore/assets/1379660/b5b32cba-ba1a-4875-8ddd-f6c7f8a1c54e)

**I assume that indexing was not implemented for terms within a taxonomy/document.**

Any ideas?

Thanks,
Dieter

**Comment by dpomt** [comment]
Any ideas on this?
